src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/MarshallingAttributeInfo.cs

```diff
@@ -820,6 +767,45 @@ private CountInfo CreateCountInfo(AttributeData marshalUsingData, ImmutableHashS
             return false;
         }
 
+        private MarshallingInfo CreateArrayMarshallingInfo(
+            ITypeSymbol elementType,
+            CountInfo countInfo,
+            MarshallingInfo elementMarshallingInfo)
+        {
+            INamedTypeSymbol? arrayMarshaller;
+            if (elementType is IPointerTypeSymbol { PointedAtType: ITypeSymbol pointedAt })
+            {
+                arrayMarshaller = _compilation.GetTypeByMetadataName(TypeNames.System_Runtime_InteropServices_GeneratedMarshalling_PtrArrayMarshaller_Metadata)?.Construct(pointedAt);
+            }
+            else
+            {
+                arrayMarshaller = _compilation.GetTypeByMetadataName(TypeNames.System_Runtime_InteropServices_GeneratedMarshalling_ArrayMarshaller_Metadata)?.Construct(elementType);
+            }
+
+            if (arrayMarshaller is null)
+            {
+                // If the array marshaler type is not available, then we cannot marshal arrays but indicate it is missing.
+                return new MissingSupportCollectionMarshallingInfo(countInfo, elementMarshallingInfo);
+            }
+
+            var (_, _, customTypeMarshallerData) = ManualTypeMarshallingHelper.GetMarshallerShapeInfo(arrayMarshaller);
+            Debug.Assert(customTypeMarshallerData is not null);
```

Comment:
I'm confused here. I see uses of `?` so the nullable feature is enabled. I see the assert, but even with that I see it used without a `null` check or use of the forgiveness operator (`!`). Does the assert remove the need for the forgiveness operator?

Reply:
`customTypeMarshallerData` is a `Nullable<CustomTypeMarshallerData>`, which is a value type. That's why we don't need the `!` operator.

Reply:
Ah. That is slightly confusing, but I get the reasoning. Thanks.